lib/Transforms/Speculation/HandshakeSpeculation.cpp

```diff
@@ -0,0 +1,101 @@
+//===- HandshakeSpeculation.cpp - Speculative Dataflows ---------*- C++ -*-===//
+//
+// Dynamatic is under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// Placement of Speculation components to enable speculative execution.
+//
+//===----------------------------------------------------------------------===//
+
+#include "dynamatic/Transforms/Speculation/HandshakeSpeculation.h"
+#include "circt/Dialect/Handshake/HandshakeOps.h"
+#include "dynamatic/Analysis/NameAnalysis.h"
+#include "dynamatic/Support/Logging.h"
+#include "dynamatic/Support/LogicBB.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/IR/OperationSupport.h"
+#include "mlir/Pass/PassManager.h"
+#include "llvm/ADT/DenseSet.h"
+#include <fstream>
+#include <string>
+#include <unordered_set>
+#include <vector>
+
+using namespace llvm::sys;
+using namespace circt;
+using namespace circt::handshake;
+using namespace mlir;
+using namespace dynamatic;
+using namespace dynamatic::speculation;
+
+HandshakeSpeculationPass::HandshakeSpeculationPass(std::string unitPositions) {
+  this->unitPositions = unitPositions;
+}
+
+template <typename T>
+LogicalResult HandshakeSpeculationPass::placeUnits(Value ctrlSignal) {
+  MLIRContext *ctx = &getContext();
+  OpBuilder builder(ctx);
+
+  for (const OpPlacement p : placements.getPlacements<T>()) {
+    // Create and connect the new Operation
+    builder.setInsertionPoint(p.dstOp);
+    T newOp = builder.create<T>(p.dstOp->getLoc(), p.srcOpResult, ctrlSignal);
+
+    // Connect the new Operation to dstOp
+    p.srcOpResult.replaceAllUsesExcept(newOp.getResult(), newOp);
+  }
+
+  return success();
+}
+
+LogicalResult HandshakeSpeculationPass::placeSpeculator() {
+  MLIRContext *ctx = &getContext();
+
+  OpPlacement place = placements.getSpeculatorPlacement();
+
+  OpBuilder builder(ctx);
+  builder.setInsertionPoint(place.dstOp);
+
+  specOp = builder.create<handshake::SpeculatorOp>(place.dstOp->getLoc(),
+                                                   place.srcOpResult);
+
+  // Replace uses of the orginal source operation's result with the speculator's
+  // result, except in the speculator's operands (otherwise this would create a
+  // self-loop from the speculator to the speculator)
+  place.srcOpResult.replaceAllUsesExcept(specOp->getResult(0), specOp);
+
+  return success();
+}
+
+void HandshakeSpeculationPass::runDynamaticPass() {
+  NameAnalysis &nameAnalysis = getAnalysis<NameAnalysis>();
+
+  if (failed(SpeculationPlacements::readFromJSON(
+          this->unitPositions, this->placements, nameAnalysis)))
+    return signalPassFailure();
+
+  if (failed(placeSpeculator()))
+    return signalPassFailure();
+
+  // Place Save operations
```

Comment:
I would replace this->specOp->getResult(n) with named accessor functions, its easy to type the wrong number somewhere and get a bug, and its easier if you ever change the numbers.

Reply:
I agree that named accessor functions could help a lot to improve the clarity. However, operations do not have that by default. Explicit accessor functions should be implemented in CIRCT first. I suggest to postpone that for a future improvement.

Reply:
Positional at the IR level is standard and almost always fine, it's only based on the complexity of the speculators outputs, (and routing them)  that I would wrap them with named outputs in the pass itself.
Just to avoid having to go look in another file, while trying to remember the specifics of the 3 auxiliary control signals.

Reply:
At first I thought that get<...> was a templated function. It took me a bit to see what you meant.
Done: now using the named accessor functions.